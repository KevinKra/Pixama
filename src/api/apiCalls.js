import API_KEY from "./apikey";

export const fetchPopularMovies = async () => {
  let response = await fetch(
    `https://api.themoviedb.org/3/discover/movie?certification_country=US&api_key=${API_KEY}`
  );
  const movies = await response.json();
  return movies.results;
};

//This method is not hooked up anywhere at this point
//data must be an object with keys of name, email and password
export const fetchNewUser = (url, data)  => {
  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Email already exists", response.message)
    } else {
      return response.json()
    }
  })
}

//this method is being imported and called in LoginCard component
export const fetchUser = (url, data)  => {

  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Not a valid login",response.message)
    } else {
      return response.json()
    }
  })
}

//Method to add a movie to a user's favorites, this method not hooked up to a component yet
// url in component method needs to be 'http://localhost:3000/api/users/favorites/new'
//data needs to be object like this: {movie_id: 550, user_id: 1, title: "Fight Club", poster_path: "/adw6Lq9FiC9zjYEpOqfq03ituwp.jpg" , release_date: "1999-10-15", vote_average:8.4, overview: "A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground \"fight clubs\" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion."}
//response will be the new favorite ID
export const postFavorite = (url, data)  => {
console.log(data)
return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Add error",response.message)
    } else {
      return response.json()
    }
  })
};


export const fetchFavorites = (url)  => {

return fetch(url)
  .then(response=> {
    if(!response.ok) {
      throw Error ("Couldn't get favorites",response.message)
    } else {
      return response.json()
    }
  })
};