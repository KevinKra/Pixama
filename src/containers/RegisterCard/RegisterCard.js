import React, { Component } from "react";
import PropTypes from "prop-types"; 
import { connect } from "react-redux";
import { loginUser } from "../../actions";
import { fetchUser, fetchNewUser } from "../../api/apiCalls";
import { Redirect } from "react-router-dom";
import { curatedData } from "../../_assets/curatedHeroData";

export class RegisterCard extends Component {
  state = {
    email: "",
    password: "",
    name: "",
    error: "",
    redirect: false
  };

  handleChange = e => {
    this.setState({
      [e.target.name]: e.target.value
    });
  };

  onSubmit = async () => {
    let newUserData = {
      name: this.state.name,
      email: this.state.email,
      password: this.state.password
    };
    let userData = { email: this.state.email, password: this.state.password };
    const newUserUrl = "http://localhost:3000/api/users/new";
    const userUrl = "http://localhost:3000/api/users";

    try {
      await fetchNewUser(newUserUrl, newUserData);
      const user = await fetchUser(userUrl, userData);
      this.props.loginUser(user.data);
      this.setState({ redirect: true });
    } catch (error) {
      this.setState({ error: error.message });
    }
    this.clearForm();
  };

  clearForm = () => {
    this.setState({
      email: "",
      password: "",
      name: ""
    });
  };

  renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to="/" />;
    }
  };

  render() {
    const divStyle = (image, position) => {
      return {
        backgroundImage: `url(${image})`,
        backgroundPosition: position,
        backgroundSize: "cover",
        backgroundRepeat: "no-repeat"
      };
    };
    const movie = curatedData[4];
    return (
      <section className="LoginContent">
        <div
          className="hero-image"
          style={divStyle(
            `https://image.tmdb.org/t/p/original${movie.backdrop_path}`,
            "center top"
          )}
        />
        <div className="opacity-filter" />
        <form className="login-card">
          <h2 className="login-card-title">Register an Account</h2>
          {this.renderRedirect()}
          <input
            className="register-name-input"
            onChange={this.handleChange}
            name="name"
            value={this.state.name}
            type="text"
            placeholder="Name"
            autoComplete="off"
            required="true"
          />
          <input
            onChange={this.handleChange}
            name="email"
            value={this.state.email}
            type="email"
            placeholder="Email"
            autoComplete="off"
            required="true"
          />
          <input
            onChange={this.handleChange}
            name="password"
            value={this.state.password}
            type="password"
            placeholder="Password"
            autoComplete="off"
            required="true"
          />
          {this.state.error && (
            <p className="login-error-text">
              {this.state.error}. Please try again.
            </p>
          )}

          <button className="register-submit-btn" type="button" onClick={this.onSubmit}>

            Register
          </button>
        </form>
      </section>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  loginUser: user => dispatch(loginUser(user))
});

RegisterCard.propTypes = {
  loginUser: PropTypes.func 
}

export default connect(
  null,
  mapDispatchToProps
)(RegisterCard);
