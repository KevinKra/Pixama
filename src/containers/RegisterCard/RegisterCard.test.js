import React from "react";
import { RegisterCard } from "./RegisterCard";
import { shallow } from "enzyme";
import * as apiCalls from '../../api/apiCalls';
import { Redirect } from "react-router-dom";

describe('RegisterCard', () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<RegisterCard loginUser={jest.fn()} />);
    instance = wrapper.instance();
  });

  it("should match a snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should have default state", () => {
    const mockState = {
      email: "",
      password: "",
      name: "",
      error: "",
      redirect: false
    }
    expect(instance.state).toEqual(mockState);
  });

  it ('should call handleChange and update state', () => {
    const mockEvent = {
      target: {
        name: "name",
        value: "Steve"
      }
    };
    wrapper.find(".register-name-input").simulate('change', mockEvent);
    expect(instance.state.name).toEqual("Steve");
  });

  it.skip('should call fetchNewUser with the correct url', async () => {
    // jest.spyOn(apiCalls, 'fetchNewUser')
    apiCalls.fetchNewUser = jest.fn()
    const newUserData = {
      name: "Taylor",
      email: "tman2272@aol.com",
      password: "password",
    };
    const url = "http://localhost:3000/api/users/new";

    wrapper.find(".register-submit-btn").simulate("click");
    await expect(apiCalls.fetchUser).toHaveBeenCalledWith(url, newUserData);
  });

  it ('should clear form', () => {
    instance.setState({
      email: "tman2272@aol.com",
      password: "password",
      name: "Taylor",
    });

    const expected = {
      email: "",
      error: "",
      password: "",
      name: "",
      redirect: false
    };

    instance.clearForm();
    expect(instance.state).toEqual(expected);
  });

  it ('should return a redirect', () => {
    instance.setState({ redirect: true });
    const expected = <Redirect to="/" />;
    const result = instance.renderRedirect();
    expect(result).toEqual(expected);
  });

  it ('should render error if there is one', () => {
    instance.setState({ error: "test is not passing"});

    expect(wrapper).toMatchSnapshot();
  });
  
});
