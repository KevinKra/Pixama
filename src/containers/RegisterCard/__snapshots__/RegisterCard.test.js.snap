// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterCard should match a snapshot 1`] = `
<section
  className="LoginContent"
>
  <div
    className="hero-image"
    style={
      Object {
        "backgroundImage": "url(https://image.tmdb.org/t/p/original/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg)",
        "backgroundPosition": "center top",
        "backgroundRepeat": "no-repeat",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="opacity-filter"
  />
  <form
    className="login-card"
  >
    <h2
      className="login-card-title"
    >
      Register an Account
    </h2>
    <input
      autoComplete="off"
      className="register-name-input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required="true"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required="true"
      type="email"
      value=""
    />
    <input
      autoComplete="off"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required="true"
      type="password"
      value=""
    />
    <button
      className="register-submit-btn"
      onClick={[Function]}
      type="button"
    >
      Register
    </button>
  </form>
</section>
`;

exports[`RegisterCard should render error if there is one 1`] = `
<section
  className="LoginContent"
>
  <div
    className="hero-image"
    style={
      Object {
        "backgroundImage": "url(https://image.tmdb.org/t/p/original/7RyHsO4yDXtBv1zUU3mTpHeQ0d5.jpg)",
        "backgroundPosition": "center top",
        "backgroundRepeat": "no-repeat",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="opacity-filter"
  />
  <form
    className="login-card"
  >
    <h2
      className="login-card-title"
    >
      Register an Account
    </h2>
    <input
      autoComplete="off"
      className="register-name-input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required="true"
      type="text"
      value=""
    />
    <input
      autoComplete="off"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required="true"
      type="email"
      value=""
    />
    <input
      autoComplete="off"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required="true"
      type="password"
      value=""
    />
    <p
      className="login-error-text"
    >
      test is not passing
      . Please try again.
    </p>
    <button
      className="register-submit-btn"
      onClick={[Function]}
      type="button"
    >
      Register
    </button>
  </form>
</section>
`;
